File: 2016/17/solution.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := string(data)
	fmt.Println(solve(input))
}

type State struct {
	pos   [2]int
	route string
}

type Direction struct {
	diff [2]int
	char string
}

func doorsState(passcode string, route string) []bool {
	hash := md5.Sum([]byte(passcode + route))
	hash2 := hex.EncodeToString(hash[:])
	doors := []bool{}
	for _, r := range hash2[:4] {
		doors = append(doors, testDoor(r))
	}
	return doors
}

func testDoor(r rune) bool {
	return r == 'b' || r == 'c' || r == 'd' || r == 'e' || r == 'f'
}

func solve(passcode string) (string, int) {
	start := State{
		pos:   [2]int{0, 0},
		route: "",
	}
	queue := []State{start}
	directions := []Direction{
		{
			diff: [2]int{0, -1},
			char: "U",
		},
		{
			diff: [2]int{0, 1},
			char: "D",
		},
		{
			diff: [2]int{-1, 0},
			char: "L",
		},
		{
			diff: [2]int{1, 0},
			char: "R",
		},
	}
	exit := [2]int{3, 3}
	shortest := ""
	longest := 0
	for i := 0; i < len(queue); i++ {
		state := queue[i]
		doors := doorsState(passcode, state.route)
		for j := 0; j < 4; j++ {
			if !doors[j] {
				continue
			}
			nextPos := [2]int{state.pos[0] + directions[j].diff[0], state.pos[1] + directions[j].diff[1]}
			if nextPos[0] < 0 || nextPos[1] < 0 || nextPos[0] > 3 || nextPos[1] > 3 {
				continue
			}
			nextState := State{
				pos:   nextPos,
				route: state.route + directions[j].char,
			}
			if nextPos == exit {
				if shortest == "" {
					shortest = nextState.route
				}
				longest = len(nextState.route)
			} else {
				queue = append(queue, nextState)
			}
		}
	}
	return shortest, longest
}

File: 2016/17/solution.ts
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";
import MinHeap from "../../utils/MinHeap";

const puzzleInput = fs.readFileSync(path.join(__dirname, "input.txt"), "utf-8");

const RMAX = 4 - 1;
const CMAX = 4 - 1;
const directions: [string, [number, number]][] = [
  ["U", [-1, 0]],
  ["D", [1, 0]],
  ["L", [0, -1]],
  ["R", [0, 1]],
];

const openRegex = /[bcdef]/;
function openDoors(s: string): boolean[] {
  return crypto
    .createHash("md5")
    .update(s)
    .digest("hex")
    .slice(0, 4)
    .split("")
    .map((a) => openRegex.test(a));
}

function heuristic(r: number, c: number): number {
  return Math.abs(r - RMAX) + Math.abs(c - CMAX);
}

function solve(puzzleInput: string): [string, number] {
  type QueueEntry = [number, number, string, number, number];
  const queue: QueueEntry[] = [[0, 0, "", 0, 0]];
  let minRoute = "";
  let maxSteps = 0;

  while (queue.length) {
    let [_, steps, route, r, c] = MinHeap.pop(queue) as QueueEntry;

    if (r === RMAX && c === CMAX) {
      if (!minRoute) minRoute = route;
      if (steps > maxSteps) maxSteps = steps;
      continue;
    }

    steps++;
    const doors = openDoors(puzzleInput + route);
    for (let i = 0; i < doors.length; i++) {
      if (!doors[i]) continue;
      const [dir, [dr, dc]] = directions[i];
      const [r2, c2] = [r + dr, c + dc];
      if (r2 < 0 || c2 < 0 || r2 > RMAX || c2 > CMAX) continue;
      MinHeap.push(queue, [
        heuristic(r2, c2) + steps,
        steps,
        route + dir,
        r2,
        c2,
      ]);
    }
  }
  return [minRoute, maxSteps];
}

const [part1, part2] = solve(puzzleInput);
console.log(part1, part2);
